### PR TITLE
Fix signin for error handling.

### PR DESCRIPTION
**Pull request by jicailiu**
1. Singin refactoring for error handling.
2. Add unittest project for c# runtime dll.
3. Add sign in unittests.

### CODE REVIEW

File: Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    using Windows.System;
+    using Windows.Security.Credentials;
+    using Windows.Security.Authentication.Web.Core;
+    using global::System.Threading.Tasks;
+
+    internal class AccountProvider
+    {
+        private WebAccountProvider provider;
+
+        public bool HasProvider() { return provider != null; }
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs]: Make this a property not a function.

File: Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        public async Task InitializeProvider(User user)
+        {
+            if (provider!=null)
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs]: HasProvider

File: Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            else
+            {
+                provider = await WebAuthenticationCoreManager.FindAccountProviderAsync("https://xsts.auth.xboxlive.com", string.Empty, user);
+            }
+        }
+
+        public virtual async Task<TokenRequestResult> RequestTokenAsync(WebTokenRequest request)
+        {
+            var result = await WebAuthenticationCoreManager.RequestTokenAsync(request);
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs]: Don't use var here.

File: Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+        public virtual async Task<TokenRequestResult> GetTokenSilentlyAsync(WebTokenRequest request)
+        {
+            var result = await WebAuthenticationCoreManager.GetTokenSilentlyAsync(request);
+            return new TokenRequestResult(result);
+        }
+
+        public WebTokenRequest CreateWebTokenRequest()
+        {
+            if (this.provider == null)
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs]: HasProvider

File: Source/Microsoft.Xbox.Services.UWP.CSharp/XboxLiveUser.cs [hunk 5]
```diff
@@ -36,13 +36,32 @@ public Task RefreshToken()
...
-        public Windows.System.User SystemUser
+        public Windows.System.User WindowsSystemUser
         {
             get
             {
                 return this.userImpl.CreationContext;
             }
         }
 
+        internal UserImpl GetImpl()
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.CSharp/XboxLiveUser.cs]: This should be a property.

File: Source/Microsoft.Xbox.Services.UWP.UnitTests/UserTest.cs [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+        [TestCategory("XboxLiveUser")]
+        [TestMethod]
+        public async Task UserSigninSilentlySuccess()
+        {
+            var user = new XboxLiveUser();
+            Assert.IsFalse(user.IsSignedIn);
+
+            AutoResetEvent signinEvent = new AutoResetEvent(false);
+            XboxLiveUser.SignInCompleted += ((Object o, SignInCompletedEventArgs args) =>
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.UnitTests/UserTest.cs]: Nit: redundant parenthesese around the delegate.

File: Source/Microsoft.Xbox.Services.UWP.UnitTests/UserTest.cs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            Assert.AreEqual(silentResult.Status, SignInStatus.Success);
+
+            Assert.IsTrue(user.IsSignedIn);
+            Assert.AreEqual(user.Gamertag, mockGamerTag);
+            Assert.AreEqual(user.XboxUserId, mockXuid);
+            Assert.AreEqual(user.AgeGroup, mockAgeGroup);
+            Assert.AreEqual(user.Privileges, mockPrivileges);
+            Assert.AreEqual(user.WebAccountId, mockWebAccountId);
+
+            Assert.IsTrue(signinEvent.WaitOne(10000), "wait signin event time out");
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.UnitTests/UserTest.cs]: You could probably make this wait like 10 ms.

File: Source/Microsoft.Xbox.Services.UWP.UnitTests/UserTest.cs [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+            AutoResetEvent signinEvent = new AutoResetEvent(false);
+            XboxLiveUser.SignInCompleted += ((Object o, SignInCompletedEventArgs args) =>
+            {
+                Assert.AreEqual(args.User, user);
+                signinEvent.Set();
+            });
+
+            var response = CreateSuccessTokenResponse();
+            user.GetImpl().Provider = CreateMockAccountProvider(null, response).Object;
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.UnitTests/UserTest.cs]: If the only reason to call GetImpl() is to get the provider, you should just create a property to return the Provider directly.  Alternatively, you could create a subclass in the test project that allows you to pass in the Provider directly.

File: Source/Microsoft.Xbox.Services.UWP.CSharp/System/TokenRequestResult.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        public string WebAccountId { get; set; }
+
+        public TokenRequestResult(WebTokenRequestResult result)
+        {
+            if (result != null)
+            {
+                this.ResponseStatus = result.ResponseStatus;
+                this.ResponseError = result.ResponseError;
+
+                if (result.ResponseData != null && result.ResponseData.Count > 0)
```
Comment (veleek) [inline Source/Microsoft.Xbox.Services.UWP.CSharp/System/TokenRequestResult.cs]: You could be fancy and say
```
var responseData = result.ResponseData?.FirstOrDefault();

if(responseData != null)
{
   ...
}

File: Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs (hunk 3 above)
Comment (jicailiu) [inline Source/Microsoft.Xbox.Services.UWP.CSharp/System/AccountProvider.cs]: why?

---
In reply to: [112354642](https://github.com/Microsoft/xbox-live-api-csharp/pull/71#discussion_r112354642) [](ancestors = 112354642)

File: Source/Microsoft.Xbox.Services.UWP.UnitTests/UserTest.cs (hunk 8 above)
Comment (jicailiu) [inline Source/Microsoft.Xbox.Services.UWP.UnitTests/UserTest.cs]: I will need to expose more test hook in userimpl when I start to implement token validation check, this way gives me more flexibility.

---
In reply to: [112355670](https://github.com/Microsoft/xbox-live-api-csharp/pull/71#discussion_r112355670) [](ancestors = 112355670)

**Pull request merged**